### PR TITLE
Reducedimbalance

### PR DESCRIPTION
**Pull request by sbird**
PR to remove a sort in the treewalk export and thus maybe reduce the tree imbalance.

### CODE REVIEW

**Comment by sbird** [comment]
This works in that it reduces wait time (and does not crash) but the extra memory accesses increase compute time more than the wait is reduced, so it is not worth it.

**Comment by rainwoodman** [comment]
In case you still have the numbers -- reducing wait time from ? to ?

**Comment by sbird** [comment]
This was only a small run so the absolute numbers are not meaningful. To ensure the export buffer filled up I gave it a max size of 20MB. Step 63 with this PR was:

SPH                              99.48 36.8%        0.21 44.4%        0.21       0.21
  Density                          61.44 22.7%        0.12 24.8%        0.12       0.12
   Comm@density.c:0375               0.86  0.3%        0.00  0.1%        0.00       0.00
   Compute@density.c:0373           46.06 17.0%        0.07 13.9%        0.06       0.07
   Init@density.c:0271               3.07  1.1%        0.04  9.0%        0.04       0.04
   Misc@density.c:0376               0.69  0.3%        0.00  0.2%        0.00       0.00
   Wait@density.c:0374              10.76  4.0%        0.01  1.6%        0.00       0.02
  Hydro                            38.04 14.1%        0.09 19.6%        0.09       0.09
   Comm@hydra.c:0200                 0.58  0.2%        0.00  0.2%        0.00       0.00
   Compute@hydra.c:0198             30.98 11.5%        0.09 17.9%        0.08       0.09
   Misc@hydra.c:0201                 0.22  0.1%        0.00  0.1%        0.00       0.00
   Wait@hydra.c:0199                 6.26  2.3%        0.01  1.5%        0.00       0.02
 Tree                            132.73 49.1%        0.15 31.8%        0.15       0.15
  Wait1@gravshort-tree.c:0166       22.86  8.5%        0.01  1.5%        0.00       0.01
  Wait2@gravshort-tree.c:0167        0.08  0.0%        0.00  0.0%        0.00       0.00
  Walk1@gravshort-tree.c:0161       96.58 35.7%        0.03  6.0%        0.02       0.04

and with the guidedtree was:
 SPH                              97.90 36.6%        0.20 43.1%        0.20       0.20
  Density                          60.64 22.7%        0.11 23.9%        0.11       0.11
   Comm@density.c:0375               1.00  0.4%        0.00  0.1%        0.00       0.00
   Compute@density.c:0373           45.99 17.2%        0.06 13.7%        0.06       0.07
   Init@density.c:0271               3.07  1.1%        0.04  9.1%        0.04       0.05
   Misc@density.c:0376               0.73  0.3%        0.00  0.2%        0.00       0.00
   Wait@density.c:0374               9.85  3.7%        0.00  0.8%        0.00       0.01
  Hydro                            37.26 13.9%        0.09 19.2%        0.09       0.09
   Comm@hydra.c:0200                 0.70  0.3%        0.00  0.1%        0.00       0.00
   Compute@hydra.c:0198             30.17 11.3%        0.08 17.6%        0.08       0.09
   Misc@hydra.c:0201                 0.26  0.1%        0.00  0.1%        0.00       0.00
   Wait@hydra.c:0199                 6.14  2.3%        0.01  1.4%        0.00       0.01
Tree                            131.62 49.2%        0.15 32.4%        0.15       0.15
 Wait1@gravshort-tree.c:0166       26.95 10.1%        0.00  1.0%        0.00       0.01
  Wait2@gravshort-tree.c:0167        0.05  0.0%        0.00  0.0%        0.00       0.00
  Walk1@gravshort-tree.c:0161       91.18 34.1%        0.03  6.1%        0.02       0.03

File: libgadget/treewalk.c [hunk 1]
```diff
@@ -707,24 +632,71 @@ ev_communicate(void * sendbuf, void * recvbuf, size_t elsize, const struct SendR
...
     char * sendbuf = mymalloc("EvDataIn", tw->Nexport * tw->query_type_elsize);
 
     tstart = second();
-    /* prepare particle data for export */
-#pragma omp parallel for
+    int * localsendcount = ta_malloc("localsendcount", int, NTask);
+    memset(localsendcount, 0, sizeof(int)*NTask);
+    /* prepare particle data for export, building the sendbuffer.
+     * Set DataIndexTable.IndexGet to the index in the send buffer. */
     for(j = 0; j < tw->Nexport; j++)
```
Comment (sbird) [inline libgadget/treewalk.c:679]: I think a lot of the pessimization comes from this loop no longer being omp parallel

**Comment by sbird** [comment]
What do you think of the current overall guidedtree implementation? It reduces wait times as long as the buffer doesn't fill up.

**Pull request closed without merging**